### PR TITLE
Implement missing HashMap methods

### PR DESCRIPTION
**Pull request by patrox**
Tests underway ...

### CODE REVIEW

File: src/main/java/javaslang/collection/HashMap.java [hunk 1]
```diff
@@ -144,17 +145,22 @@ public boolean containsValue(V value) {
...
-        throw new UnsupportedOperationException("TODO");
+        return tree.iterator().map(Entry::of).toSet();
     }
 
     @Override
     public HashMap<K, V> filter(Predicate<? super Entry<K, V>> predicate) {
-        throw new UnsupportedOperationException("TODO");
+        Objects.requireNonNull(predicate, "predicate is null");
+        return foldLeft(HashMap.<K, V>empty(), (acc, entry) -> {
+            if (predicate.test(entry)) return acc.put(entry);
```
Comment (danieldietrich) [inline src/main/java/javaslang/collection/HashMap.java]: Let's do curly braces `if { ... } else { ... }`. It's less error prone. Thx! :)

Comment (patrox) [inline src/main/java/javaslang/collection/HashMap.java]: yes, you're right - I missed that!

**Comment by danieldietrich** [comment]
Nice stuff, Patryk!

**Comment by danieldietrich** [comment]
@ruslansennov @patrox good night!

**Comment by patrox** [comment]
Good night!

File: src/main/java/javaslang/collection/HashMap.java (hunk 1 above)
Comment (danieldietrich) [inline src/main/java/javaslang/collection/HashMap.java]: I will merge it - it's working perfectly - braces can be added later...

**Pull request merged**